rust/relay/src/server.rs

```diff
@@ -955,6 +964,60 @@ where
 
         tracing::info!(target: "relay", channel = %chan.value(), %client, %peer, %allocation, "Channel binding is now deleted (and can be rebound)");
     }
+
+    fn make_error_response(
+        &mut self,
+        error_code: impl Into<ErrorCode>,
+        request: &impl StunRequest,
+        error_level: ResponseErrorLevel,
+    ) -> Message<Attribute> {
+        let error_code = error_code.into();
+
+        error_level.log(request.method(), error_code.reason_phrase());
+
+        let mut message = Message::new(
+            MessageClass::ErrorResponse,
+            request.method(),
+            request.transaction_id(),
+        );
+
+        let is_auth_error = error_code == ErrorCode::from(Unauthorized)
+            || error_code == ErrorCode::from(StaleNonce);
+
+        message.add_attribute(Attribute::from(error_code));
+
+        // In case of a 401 or 438 response, attach a realm and nonce.
+        if is_auth_error {
+            let new_nonce = Uuid::from_u128(self.rng.gen());
+
+            self.add_nonce(new_nonce);
+
+            message.add_attribute(Nonce::new(new_nonce.to_string()).unwrap());
+            message.add_attribute((*FIREZONE).clone());
+        }
+
+        message
+    }
+}
+
+enum ResponseErrorLevel {
+    Warn,
+    Debug,
+}
```

Comment:
I had a similar design at some point :)

The question I asked myself then was: Why am I creating a new `Level` enum when there is already `tracing::Level`? But if you use `tracing::Level`, you need to handle all cases which goes against YAGNI.

Plus, the additional parameter creates a more verbose formatting in many cases. That is how I ended up with two functions because I thought, the duplication isn't that bad.

Reply:
> The question I asked myself then was: Why am I creating a new Level enum when there is already tracing::Level? But if you use tracing::Level, you need to handle all cases which goes against YAGNI.

I don't think it's so weird to have this type different from `Level`. It's just that `make_error_response`'s domain, disregarding the other parameters, is `ResponseErrorLevel` which is a subset of `Level` (more like isomorphic to a subset of `Level`), so it makes sense to define a separate type to represent this.

> Plus, the additional parameter creates a more verbose formatting in many cases. That is how I ended up with two functions because I thought, the duplication isn't that bad.

I don't mind the duplication but it just feels a bit weird to have 2 functions which only differ on the log level. though I don't feel too strongly about this. But I think, for someone w/o context, that having to "fill up a parameter" makes you think more about what you are going to put there compared to using one of the 2 functions, as someone might miss that the other one exists.

Reply:
Happy to go with this if we inline the `log` function :)